### PR TITLE
apptainer: 1.2.2 -> 1.2.4, singularity-ce: 3.11.4 -> 4.0.1

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Things done

Changelogs:

- `apptainer`: https://github.com/apptainer/apptainer/releases/tag/v1.2.4
- `singularity`: https://github.com/sylabs/singularity/releases/tag/v4.0.1

cc @jbedo @ShamrockLee 

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by matthiasbeyer — APPROVED** [review]
Diff LGTM

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 266628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>apptainer</li>
    <li>apptainer-overriden-nixos</li>
    <li>singularity</li>
    <li>singularity-overriden-nixos</li>
    <li>udocker</li>
    <li>udocker.dist</li>
  </ul>
</details>

**Comment by mweinelt** [comment]
Sorry, no idea what this is. Maybe Matthias can merge, since they approved the PR already.

**Comment by GaetanLepage** [comment]
> Sorry, no idea what this is. Maybe Matthias can merge, since they approved the PR already.

Ok no worry ! Sorry for uselessly pinging you then

**Comment by ShamrockLee** [comment]
I'll test the image-running functionality this weekends. (Apptainer doesn't build on NixOnDroid, and I need my laptop to test it.)

The test case already cover the image building. The binary is proved to build images once the `passthru.tests.image-hello-cowsay` package test has passed.

**Comment by ShamrockLee** [comment]
Also, there's a major version bump for Sylabs's SingularityCE, which brings full support to building and running OCI images instead of the native Apptainer/Singularity SIF images. Has anyone already tested it?

It's also a strong signal that we should prefer definition files to sandboxes or direct mksquasfs when implementing `singularity-tools.buildImage`.

**Review by ShamrockLee — APPROVED** [review]

**Comment by ShamrockLee** [comment]
Tested running the images built as `apptainer.passthru.tests.image-hello-cowsay` and `singularity.passthru.tests.image-hello-cowsay`.

File: pkgs/applications/virtualization/singularity/packages.nix [hunk 1]
```diff
@@ -7,14 +7,14 @@ let
...
-      version = "1.2.2";
+      version = "1.2.4";
       projectName = "apptainer";
 
       src = fetchFromGitHub {
         owner = "apptainer";
         repo = "apptainer";
-        rev = "v${version}";
-        hash = "sha256-CpNuoG+QykP+HDCyFuIbZKYez5XnYrE75SWFoWu34rg=";
+        rev = "refs/tags/v${version}";
```
Comment (ShamrockLee) [inline pkgs/applications/virtualization/singularity/packages.nix:16]: Sorry for the noise, but why do we need the `refs/tags/` prefix?

I just noticed this when backporting an Apptainer security update (#319649).

Comment (GaetanLepage) [reply pkgs/applications/virtualization/singularity/packages.nix:16]: We don't "need" it per say, but it is recommended as it prevents any branch/tag ambiguity.

**Pull request merged**